tests/test_collection.py

```diff
@@ -0,0 +1,366 @@
+        loaded = schema.load(data_in)
+        self.assertEqual(
+            loaded,
+            AnySequence(value=(1, "2", 3.0, True, None, {"this", "is a", "set"})),
+        )
+        self.assertEqual(schema.dump(loaded), data_in)
+
+    def test_optional_sequence(self):
+        @dataclass
+        class IntOptionalSequence:
+            value: Optional[Sequence[int]]
+
+        schema = IntOptionalSequence.Schema()
+        for data_in in ({"value": list(range(5))}, {"value": []}, {"value": None}):
+            self.assertEqual(schema.dump(schema.load(data_in)), data_in)
+
+    def test_sequence_of_optional(self):
+        @dataclass
+        class OptionalIntSequence:
+            value: Sequence[Optional[int]]
+
+        schema = OptionalIntSequence.Schema()
+        for data_in in ({"value": list(range(5))}, {"value": [1, None, 2, None]}):
+            self.assertEqual(schema.dump(schema.load(data_in)), data_in)
+
+    def test_nested_sequence(self):
+        @dataclass
+        class SequenceIntSequence:
+            value: Sequence[Sequence[int]]
+
+        schema = SequenceIntSequence.Schema()
+        for data_in in (
+            {"value": []},
+            {"value": [[], []]},
+            {"value": [[1, 2, 3], [4, 5, 6]]},
+        ):
+            self.assertEqual(schema.dump(schema.load(data_in)), data_in)
+
+    def test_sequence_of_dataclass(self):
+        @dataclass
+        class Elm:
+            value: int
+
+        @dataclass
+        class SequenceIntSequence:
+            value: Sequence[Elm]
+
+        schema = SequenceIntSequence.Schema()
+        for data_in in (
+            {"value": []},
+            {"value": [{"value": 1}]},
+        ):
+            self.assertEqual(schema.dump(schema.load(data_in)), data_in)
+
+    def test_sequence_of_optional_dataclass(self):
+        @dataclass
+        class Elm:
+            value: int
+
+        @dataclass
+        class SequenceIntSequence:
+            value: Sequence[Optional[Elm]]
+
+        schema = SequenceIntSequence.Schema()
+        for data_in in (
+            {"value": []},
+            {"value": [None]},
+            {"value": [{"value": 1}]},
+        ):
+            self.assertEqual(schema.dump(schema.load(data_in)), data_in)
+
+    def test_serializing_sets_as_sequence_is_possible_but_risky(self):
+        @dataclass
+        class StrSequence:
+            value: Sequence[str]
+
+        str_schema = StrSequence.Schema()
+
+        # you can dump it, but order will be random
+        sequence = (
+            "this",
+            "has",
+            "only",
+            "1",
+            "chance",
+            "in",
+            "40M",
+            "to",
+            "keep",
+            "it's",
+            "ordered",
+        )
+        dump = str_schema.dump(StrSequence(value=set(sequence)))
+        self.assertNotEqual(
+            dump,
+            {"value": sequence},
+        )
```

Comment:
You shouldn't assertNotEqual here. You should probably assert that `dump["value"] == set(sequence)` instead

Reply:
Agreed. Not a very deterministic test as it is. I wanted to point out the fact that dumping set will yield a random sorted list, but I guest that checking that is out of scope.